ddtrace/internal/module.py

```diff
@@ -189,7 +193,25 @@ def _create_module(self, spec):
         return self.loader.create_module(spec)
 
     def _exec_module(self, module):
-        self.loader.exec_module(module)
+        # Collect and run only the first hook that matches the module.
+        pre_exec_hook = None
+
+        for _ in sys.meta_path:
+            if isinstance(_, ModuleWatchdog) and not pre_exec_hook:
+                try:
+                    for cond, hook in _._pre_exec_module_hooks:
+                        if isinstance(cond, str) and cond == module.__name__ or cond(module.__name__):
+                            # Several pre-exec hooks could match, we keep the first one
+                            pre_exec_hook = hook
+                            break
+                except Exception:
+                    log.debug("Exception happened while processing pre_exec_module_hooks", exc_info=True)
+
+        if pre_exec_hook:
+            pre_exec_hook(self, module)
+        else:
+            self.loader.exec_module(module)
```

Comment:
```suggestion
        for _ in sys.meta_path:
            if isinstance(_, ModuleWatchdog) and not pre_exec_hook:
                try:
                    for cond, hook in _._pre_exec_module_hooks:
                        if isinstance(cond, str) and cond == module.__name__ or cond(module.__name__):
                            # Several pre-exec hooks could match, we keep the first one
                            hook(self, module)
                            return
                except Exception:
                    log.debug("Exception happened while processing pre_exec_module_hooks", exc_info=True)
                    break

        self.loader.exec_module(module)
```

Reply:
ah right! I lost sight of the code below! 😢 then we can exit the outer loop early?
```suggestion
        pre_exec_hook = None

        for _ in sys.meta_path:
            if pre_exec_hook is not None:
                pre_exec_hoo(self, module)
                break
            if isinstance(_, ModuleWatchdog):
                try:
                    for cond, hook in _._pre_exec_module_hooks:
                        if isinstance(cond, str) and cond == module.__name__ or cond(module.__name__):
                            # Several pre-exec hooks could match, we keep the first one
                            pre_exec_hook = hook
                            break
                except Exception:
                    log.debug("Exception happened while processing pre_exec_module_hooks", exc_info=True)
        else:
            self.loader.exec_module(module)
```

Reply:
In this case we'll not execute ``self.loader.exec_module(module)`` (and we should) if there are hooks and none of them matches the ``cond`` condition.

I believe we can do:
```
        for _ in sys.meta_path:
            [...]
            if pre_exec_hook is not None:
                break
```
to exit the outer loop early, but then keep the
```
        if pre_exec_hook:
            pre_exec_hook(self, module)
        else:
            self.loader.exec_module(module)
```